File: modules/lift/src/main/scala/org/scalabox/lift/extension/LiftService.scala
package org.scalabox.lift.extension

import scala.collection.mutable.HashSet
import scala.collection.mutable.SynchronizedSet
import org.jboss.msc.service.{ServiceName, StartContext, StopContext, Service}
import java.util.concurrent.atomic.AtomicLong
import java.lang.Thread

/**
 * // TODO: Document this
 * @author Galder Zamarreño
 * @since // TODO
 */
class LiftService(suffix: String, tick: Long) extends Service[LiftService] {

   // Constructor
   currentTick.set(tick)

   // Lazy so that it can be initialized when it's used by the constructor
   lazy val currentTick = new AtomicLong(10000)

   val deployments = new HashSet[String] with SynchronizedSet[String]

   val coolDeployments = new HashSet[String] with SynchronizedSet[String]

   val output = new Thread {
      override def run: Unit = {
         while (true) {
            try {
               Thread.sleep(currentTick.get)
               print("""|Current deployments deployed while %s tracking active:
                        |Cool: %d""".format(suffix, coolDeployments.size).stripMargin)
            }
            catch {
               case e: InterruptedException => Thread.interrupted()
            }
         }
      }
   }

   override def start(context: StartContext) = output.start()

   override def stop(context: StopContext) = output.interrupt()

   override def getValue = this

   def addDeployment(name: String) = deployments.add(name)

   def addCoolDeployment(name: String) = coolDeployments.add(name)

   def removeDeployment(name: String) {
      deployments.remove(name)
      coolDeployments.remove(name)
   }

   def setTick(tick: Long) = currentTick.set(tick)

   def getTick = currentTick.get()

}

object LiftService {

   def createServiceName(suffix: String) =
      ServiceName.JBOSS.append("lift", suffix)

}

File: modules/lift/src/main/_java/org/scalabox/lift/extension/LiftService.java
package org.scalabox.lift.extension;

import org.jboss.msc.service.Service;
import org.jboss.msc.service.ServiceName;
import org.jboss.msc.service.StartContext;
import org.jboss.msc.service.StartException;
import org.jboss.msc.service.StopContext;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicLong;

/**
 * // TODO: Document this
 *
 * @author Galder Zamarreño
 * @since // TODO
 */
public class LiftService implements Service<LiftService> {

   private final AtomicLong tick = new AtomicLong(10000);

   private final Set<String> deployments =
         Collections.synchronizedSet(new HashSet<String>());

   private final Set<String> coolDeployments =
         Collections.synchronizedSet(new HashSet<String>());

   private final String suffix;

   private final Thread OUTPUT = new Thread() {
      @Override
      public void run() {
         while (true) {
            try {
               Thread.sleep(tick.get());
               System.out.println("Current deployments deployed while " + suffix + " tracking active:\n" + deployments
                                        + "\nCool: " + coolDeployments.size());
            } catch (InterruptedException e) {
               interrupted();
               break;
            }
         }
      }
   };

   public LiftService(String suffix, long tick) {
      this.suffix = suffix;
      this.tick.set(tick);
   }

   @Override
   public void start(StartContext context) throws StartException {
      OUTPUT.start();
   }

   @Override
   public void stop(StopContext context) {
      OUTPUT.interrupt();
   }

   @Override
   public LiftService getValue() throws IllegalStateException, IllegalArgumentException {
      return this;
   }

   public void addDeployment(String name) {
      deployments.add(name);
   }

   public void addCoolDeployment(String name) {
      coolDeployments.add(name);
   }

   public void removeDeployment(String name) {
      deployments.remove(name);
      coolDeployments.remove(name);
   }

   void setTick(long tick) {
      this.tick.set(tick);
   }

   public long getTick() {
      return this.tick.get();
   }

   public static ServiceName createServiceName(String suffix) {
      return ServiceName.JBOSS.append("lift", suffix);
   }

}
